Migrate env vars to YAML with overrides and env injection

mbunwe-victor opened the pull request:
This PR refactors the Keycloak SSI CLI configuration system by migrating environment variables from the .env file into a structured YAML configuration. It improves readability, maintainability, and flexibility for different environments.

# Changes Introduced

1. ### YAML-based Configuration 

    - All environment variables are now stored in config.yaml in the project root.

2. ## Override Mechanism

    - Introduces config.override.yaml to allow environment-specific or partial overrides without modifying the base configuration.

    - Environment variables still take the highest priority.

3. ## Environment Variable Injection

    - Sensitive values and dynamic references in YAML now support ${VAR_NAME} syntax for seamless injection from the shell environment.
    
    Improvement PR: https://github.com/adorsys/keycloak-ssi-deployment/pull/190

mbunwe-victor committed:
feat(config): migrate env vars to YAML with overrides and env injection

mbunwe-victor committed:
remove all reference to the .env file

mbunwe-victor committed:
updated the env variable naming in all scripts

mbunwe-victor committed:
refactor(config): update paths and environment variable references in configuration files

mbunwe-victor committed:
fix merged conflicts

mbunwe-victor committed:
update the config.yaml file

Awambeng commented on config.yaml:
```diff
@@ -0,0 +1,133 @@
+# -----------------------------------------------------------------------------
+keystore:
+  # Keycloak keystore file path
+  file: "${PROJECT_TARGET_DIR}/kc_keystore.pkcs12"
```
Please rename `file` to `path` and update all references to this variable accordingly.

Awambeng commented on src/config/realm_configs/keycloak-config-release.json:
This file appears outdated and is no longer in use. I suggest removing it entirely to avoid confusion. It will be added back in the future.

Awambeng commented on config.yaml:
```diff
@@ -0,0 +1,133 @@
+# -----------------------------------------------------------------------------
+# Keycloak URLs and endpoints
+# -----------------------------------------------------------------------------
+urls:
```
The `urls` section feels a bit awkward for variable naming. Do you think we should rename it or redistribute its values into more fitting sections?

Awambeng commented on config.yaml:
```diff
@@ -0,0 +1,133 @@
+  # HTTPS port for Keycloak
+  https_port: 8443
+  # Keycloak admin address
+  admin_addr: "https://localhost:${KEYCLOAK_HTTPS_PORT}"
```
I’m wondering if there’s a variable named `KEYCLOAK_HTTPS_PORT`. Could you please check if this works?

mbunwe-victor replied:
Yes @Awambeng, there is a method in the `helper.sh` that reads the value of ` https_port` from the `config.yaml` and then exports it as an environmental variable named `KEYCLOAK_HTTPS_PORT`
See here:
                ```
                export "$env_var_name"="$value"
        case "$key" in
            "keycloak_endpoints.https_port")
                export KEYCLOAK_HTTPS_PORT="$value"
                ;;
            "keycloak_endpoints.admin_addr")
                export KEYCLOAK_ADMIN_ADDR="$value"
                ;;
        esac
        ```

Awambeng commented on config.yaml:
The way variables are currently defined doesn’t allow referencing other variables within this file.
For example:

```
keycloak:
  target_branch: "datev/develop"
  version: "999.0.0-SNAPSHOT"
  oid4vci_dir: "keycloak_${KC_TARGET_BRANCH}"
```

Here, `oid4vci_dir` ends up as just `keycloak` because `KC_TARGET_BRANCH` is not defined, so the interpolation doesn’t work.
This issue affects all places in this file where you try to reference one variable inside another.
Please review and adjust accordingly

Awambeng reviewed (changes requested):
@mbunwe-victor Please take a look at the comments I left. Also, try deleting your `target` folder before testing to make sure everything runs as expected. I stopped the review for now since I ran into some issues while testing.

mbunwe-victor committed:
address Bolan comments

mbunwe-victor committed:
address Bolan comments

Awambeng commented on Dockerfile:
```diff
@@ -9,10 +9,15 @@ WORKDIR $WORK_DIR
 
 # Copy environment file
-COPY .env ./
+COPY config.yaml config.override.yaml ./
```
Could you please verify if we can still override variables at runtime with these new changes? Previously, we had an entrypoint script that allowed us to pass a .env file from a specific path to override variables.

Awambeng reviewed (changes requested):
Adding to this comment, could you please also address the other points we discussed earlier?

mbunwe-victor committed:
address PR comments

mbunwe-victor committed:
Merge branch 'develop' into yaml-overridable-config-541

mbunwe-victor committed:
address Bolan comments

mbunwe-victor committed:
Merge branch 'yaml-overridable-config-541' of github.com:adorsys/keycloak-ssi-deployment into yaml-overridable-config-541

mbunwe-victor commented:
Hello @Awambeng, please could you take a look again?

Awambeng commented on docker-compose.sh:
I don’t really see the importance of this script. Please let’s avoid adding unnecessary files to the project, unless there’s a clear justification for it.

mbunwe-victor replied:
I added this because of docker-compose.yaml wasn't able to load variables from the config.yaml file as you understood here: https://github.com/adorsys/keycloak-ssi-deployment/pull/185#discussion_r2534515053, but it has been removed.

Awambeng commented on scripts/docker-entrypoint.sh:
I thought we wanted to completely remove references to the .env file. I noticed that this file still uses .env values to override configuration at runtime in the container. Was this intentional?

Additionally, the scripts that were added to the new `scripts` directory should be moved to their proper locations. Creating a new folder for them disperses related files unnecessarily.

mbunwe-victor replied:
Yes, @Awambeng, that was intentional because the `docker-compose.yaml` could not directly load the `config.yaml` file.
 
 To clarify: even though Keycloak prints messages about /opt/keycloak/.env, we are not actually using a static .env file anymore.
Here’s what happens now:
- Our config.yaml (and config.override.yaml) is the real source of truth.
- The CLI generates a temporary .env.generated file from those YAML values.
- Docker Compose uses this generated file to inject the variables directly into the Keycloak container at runtime.
- Inside the container, Keycloak checks for a .env file at /opt/keycloak/.env. Since we don’t place one there, it logs “no custom .env file,” which is expected.

Although you still see .env-related log messages, no static .env file is being used, and all configuration overrides are coming from config.yaml, exactly as intended.

Awambeng commented on docs/CLI_GUIDE.md:
```diff
@@ -13,6 +13,34 @@ CLI Tool • Deployment • Configuration • Testing
+
+1.  **Configuration Files**: [`config.yaml`](config.yaml) defines the base configuration, and [`config.override.yaml`](config.override.yaml) provides optional overrides.
+2.  **[`src/utils/helper.sh`](src/utils/helper.sh)**: This script's [`setup_environment()`](src/utils/helper.sh:29) and [`load_configuration()`](src/utils/helper.sh:145) functions read and merge these YAML files. The [`export_yaml_as_env()`](src/utils/helper.sh:60) function then converts the merged YAML properties into environment variables, handling placeholder substitution in a two-pass process ...
+3.  **Loading into Other Scripts**: Other shell scripts source [`src/utils/helper.sh`](src/utils/helper.sh) and call [`init_script()`](src/utils/helper.sh:292) to make these resolved environment variables available for their execution.
```
To keep the documentation focused, please simplify this section and limit it to showing how to configure new variables.

Awambeng commented on scripts/generate-compose-env.sh:
I understand why you added this, since Docker Compose can only load variables from `.env`. However, since we’re migrating from `.env` to a YAML-based config, I suggest adding a command in our CLI to start Docker Compose. That way, our CLI helpers can handle this limitation. What do you think?

mbunwe-victor replied:
Yeah, that is a good idea.

Awambeng commented on src/deployment/0.start-kc-oid4vci.sh:
```diff
@@ -72,17 +72,22 @@ if [[ "${1:-}" == "-d" ]]; then
+# Set standard Keycloak admin environment variables for initial setup
+export KC_BOOTSTRAP_ADMIN_USERNAME="$KEYCLOAK_BOOTSTRAP_ADMIN_USERNAME"
+export KC_BOOTSTRAP_ADMIN_PASSWORD="$KEYCLOAK_BOOTSTRAP_ADMIN_PASSWORD"
+
```
I don’t see any added value from this change. Was it necessary?

mbunwe-victor replied:
I was resolving an issue I had when running the `keycloak-ssi config` command.

Awambeng commented on src/utils/helper.sh:
```diff
@@ -48,18 +49,144 @@ setup_environment() {
+    [[ -n "${ENDPOINTS_DEV_CLIENTS_TEST_CLIENT:-}" ]] && export TEST_CLIENT_URL="${ENDPOINTS_DEV_CLIENTS_TEST_CLIENT}"
+    [[ -n "${ENDPOINTS_ISSUER_BACKEND:-}" ]] && export ISSUER_BACKEND_URL="${ENDPOINTS_ISSUER_BACKEND}"
+    [[ -n "${ENDPOINTS_ISSUER_FRONTEND:-}" ]] && export ISSUER_FRONTEND_URL="${ENDPOINTS_ISSUER_FRONTEND}"
+    [[ -n "${KEYCLOAK_ENDPOINTS_ISSUER_DID:-}" ]] && export ISSUER_DID="${KEYCLOAK_ENDPOINTS_ISSUER_DID}"
```
Could you please explain why we have this? I don’t quite understand its purpose or usage

mbunwe-victor replied:
Same here, I have removed it

Awambeng commented on Dockerfile:
```diff
@@ -9,10 +9,15 @@ WORKDIR $WORK_DIR
 
 # Copy environment file
-COPY .env ./
+COPY config.yaml config.override.yaml ./
```
I still don’t understand why we need to copy `config.override.yaml` into the container. Isn’t this supposed to be a production-specific file?

mbunwe-victor replied:
I have removed it

Awambeng commented on Dockerfile:
```diff
@@ -28,25 +33,38 @@ FROM eclipse-temurin:21-jdk-jammy
+# Install gettext-base for envsubst (used by entrypoint script)
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends gettext-base && \
+    rm -rf /var/lib/apt/lists/*
```
why this?

Awambeng commented on Dockerfile:
```diff
@@ -28,25 +33,38 @@ FROM eclipse-temurin:21-jdk-jammy
+# Create directory for custom .env files (to be mounted at runtime)
+RUN mkdir -p /opt/keycloak/env && \
+    chown -R keycloak:keycloak /opt/keycloak
+
```
we donot want to have anything to do with .env anymore

Awambeng commented on Readme.md:
Please make sure to update this file after addressing all the comments.

Awambeng reviewed (changes requested):
@mbunwe-victor I left a few comments. Could you take a look when you get a chance?

mbunwe-victor committed:
address Bolan's comments

mbunwe-victor commented:
Hello @Awambeng, could you check this again?

hugoib commented on Dockerfile:
```diff
@@ -28,25 +33,24 @@ FROM eclipse-temurin:21-jdk-jammy
-# Set entrypoint
-ENTRYPOINT ["sh", "/opt/keycloak/entrypoint.sh"]
+# Set entrypoint script
+ENTRYPOINT ["/bin/sh", "-c", "cd $KC_INSTALL_DIR && exec bin/kc.sh $KC_START $KC_DB_OPTS --features=oid4vc-vci,oid4vc-vpauth"]
```
The entry point KC_INSTALL_DIR is only here in the codebase and none of the new env exports set KC_INSTALL_DIR. The scripts use KEYCLOAK_INSTALL_DIR. Runtime container likely will fail because $KC_INSTALL_DIR is empty. Please review.

hugoib commented on src/config/realm_configs/keycloak-config-dev.json:
Review https://github.com/adorsys/keycloak-ssi-deployment/blob/e5aca93645dec5965fa4af8ff972096f3b440345/src/config/realm_configs/keycloak-config-dev.json#L68 I think the other scripts export CLIENTS_SECRET and that reference needs to be updated.

hugoib commented on keycloak-ssi.sh:
```diff
@@ -231,6 +234,40 @@ cmd_stop() {
+
+    # Clean up the temporary .env file
+    log "Cleaning up temporary .env file..."
+    rm -f "$env_file"
```
Do this: 

Comment the removal of the file:
 # rm -f "$env_file"

Run:
- ./keycloak-ssi.sh compose config


You will see that .env.generated is empty... This happens because export_yaml_as_env does not outputs what it needs to output. The helper never writes anything to stdout.

mbunwe-victor committed:
address Hugo's comments

Awambeng commented:
@mbunwe-victor I’ve changed the base branch of this PR, which has caused some conflicts. If you don’t mind, we could sync tomorrow to resolve them before you address the comments Hugo left

mbunwe-victor committed:
fix comflict

mbunwe-victor committed:
address Hugo's comment

mbunwe-victor commented:
@hugoib and @Awambeng, please could you take a look again?

hugoib reviewed (approved):
I reviewed functionality. 

I uninstalled and installed the script. I can run both tests:

keycloak-ssi test preauth IdentityCredential
  keycloak-ssi test authcode IdentityCredential

Also reviewed the code. Looks good! Thanks!

Awambeng commented on src/utils/helper.sh:
```diff
@@ -33,33 +33,165 @@ setup_environment() {
+    if ! command -v yq &>/dev/null; then
+        error "yq is required for configuration management. Please install yq (version 4+)."
+        return 1
     fi
```
Quick question — do we still need to repeat this check?
There is already a method that ensures all dependencies are installed.
See this line for reference: https://github.com/adorsys/keycloak-ssi-deployment/pull/185/files#diff-3a397c7ab03a9c0207c9aa54a20c73e1fc8e69ba50e7c06dff7925f6ab4aba47R289

Awambeng commented on src/utils/helper.sh:
```diff
@@ -107,7 +239,7 @@ stop_keycloak() {
         DOCKER_COMPOSE_COMMAND="$(detect_docker_compose)"
         log "Stopping and removing database container..."
-        eval "$DOCKER_COMPOSE_COMMAND -f \"$DOCKER_COMPOSE_FILE\" down -v db" || \
+        eval "$DOCKER_COMPOSE_COMMAND -f \"$DOCKER_COMPOSE_FILE\" down -v"
```
Do you have any reasons for removing this?

Awambeng commented on src/utils/helper.sh:
```diff
@@ -172,9 +304,17 @@ init_script() {
 
+    # Export shorthand variables for convenience, as used in config.yaml
+    export TARGET_DIR="${PROJECT_TARGET_DIR}"
+    export TOOLS_DIR="${PROJECT_TOOLS_DIR}"
```
I don’t quite understand why we need to explicitly export these variables.
My understanding was that `load_configuration` is responsible for loading all variables required by the scripts.
Could you clarify why this additional export step is necessary?

mbunwe-victor replied:
These changes ensure that user_key_proof_header.json is created and accessed from the correct target directory within the project, resolving the file not found error that I got when running the command `keycloak-ssi test preauth IdentityCredential`

Awambeng commented on src/deployment/0.start-kc-oid4vci.sh:
```diff
@@ -77,17 +78,20 @@ if [[ "${1:-}" == "-d" ]]; then
   LOG_FILE="$LOG_DIR/keycloak.log"
   log "Detaching Keycloak; logs will be written to $LOG_FILE"
-  nohup bash -c "exec bin/kc.sh $KC_START $KC_DB_OPTS --features=oid4vc-vci" \
+  KEYCLOAK_ADMIN="$KEYCLOAK_BOOTSTRAP_ADMIN_USERNAME" KEYCLOAK_ADMIN_PASSWORD="$KEYCLOAK_BOOTSTRAP_ADMIN_PASSWORD" \
```
Please update these variables to use `KC_BOOTSTRAP_ADMIN_USERNAME` and `KC_BOOTSTRAP_ADMIN_PASSWORD`

mbunwe-victor replied:
Changing this back to KC instead is gonna work cause it defeats the format in the `config.yaml` file.
```
keycloak:
  # Branch to use for custom Keycloak
  target_branch: "datev/develop"
  # Keycloak version
  version: "999.0.0-SNAPSHOT"
  # Directory name for cloned Keycloak repo
  oid4vci_dir: "keycloak_${KEYCLOAK_TARGET_BRANCH}"
  # Keycloak tarball path
  tarball_path: "${PROJECT_TARGET_DIR}/keycloak-${KEYCLOAK_VERSION}.tar.gz"
  # Keycloak installation directory (after unpack)
  install_dir: "${PROJECT_TOOLS_DIR}/keycloak-${KEYCLOAK_VERSION}"
  # Keycloak repository to clone
  repo_url: "https://github.com/adorsys/keycloak-oid4vc.git"
  # Bootstrap admin credentials
  bootstrap:
    admin_username: "admin"
    admin_password: "admin"
  # Keycloak realm
  realm: "oid4vc-vci"
  ```

Awambeng commented on src/deployment/0.start-kc-oid4vci.sh:
```diff
@@ -77,17 +78,20 @@ if [[ "${1:-}" == "-d" ]]; then
   disown || true
 else
-  bash -c "exec bin/kc.sh $KC_START $KC_DB_OPTS --features=oid4vc-vci"
+  KEYCLOAK_ADMIN="$KEYCLOAK_BOOTSTRAP_ADMIN_USERNAME" KEYCLOAK_ADMIN_PASSWORD="$KEYCLOAK_BOOTSTRAP_ADMIN_PASSWORD" \
```
same here.

Awambeng replied:
same here.

Awambeng commented on src/utils/update_sdjwt_vct.sh:
this file was removed as this feature is not yet supported in keycloak

Awambeng commented on .gitignore:
```diff
@@ -7,3 +7,20 @@ target/
+terraform.tfstate
+terraform.tfstate.backup
+.terraform/
+**/.terraform.lock.hcl
```
Do we still have this files in the project? i guess you didn't handle the conflict resolution well.

Awambeng replied:
This too should be removed as the files are no longer part of the project.

Awambeng commented on Dockerfile:
```diff
@@ -28,25 +33,24 @@ FROM eclipse-temurin:21-jdk-jammy
 
 # Set working directory
 WORKDIR /opt/keycloak
+ENV JAVA_TMPDIR=/tmp
```
Is this line necessary?

Awambeng commented on keycloak-ssi.sh:
```diff
@@ -83,6 +84,8 @@ show_help() {
 "  keycloak-ssi setup" \
 "  keycloak-ssi setup -d" \
+"  keycloak-ssi compose up -d" \
+"  keycloak-ssi compose down" \
```
Do you mind stopping the Docker Compose stack with `-v` to ensure old volumes don't cause issues in subsequent runs?

Awambeng replied:
Do you mind stopping the Docker Compose stack with `-v` to ensure old volumes don't cause issues in subsequent runs?

Awambeng commented on keycloak-ssi.sh:
```diff
@@ -231,6 +234,43 @@ cmd_stop() {
+        DOCKER_COMPOSE_CMD=(docker-compose)
+    else
+        error "Neither 'docker compose' (v2) nor 'docker-compose' (v1) is installed."
+    fi
```
This logic appears identical to the helper method. Can we call that method here to avoid duplication?

Awambeng commented on docs/CLI_GUIDE.md:
The doc should reflect the additional commands you introduced — please update it accordingly.

Awambeng commented on .gitignore:
```diff
@@ -7,3 +7,20 @@ target/
+**/.terraform.lock.hcl
+
+# Docker Compose environment file (generated from config.yaml)
+.env
```
```suggestion
.env*
```
This will also ignore the temporary `.env` files.
Please ensure only a single newline remains in this file.

Awambeng commented on src/utils/helper.sh:
```diff
@@ -183,4 +323,6 @@ init_script() {
-export -f urlencode detect_docker_compose init_script ensure_directory_exists check_dependencies
+export -f urlencode detect_docker_compose init_script ensure_directory_exists check_dependencies export_yaml_as_env
+
+
```
remove this extra line

Awambeng commented on src/utils/helper.sh:
```diff
@@ -33,33 +33,165 @@ setup_environment() {
+    echo "$yaml_content" | yq '(.. | select(tag == "!!str")) |= envsubst'
 }
 
+
```
remove this empty line.

Awambeng commented on Dockerfile:
```diff
@@ -28,25 +33,24 @@ FROM eclipse-temurin:21-jdk-jammy
 # Create a non-privileged user
 RUN groupadd -r keycloak && useradd -r -g keycloak keycloak
 
+
```
remove this pls.

Awambeng commented on keycloak-ssi.sh:
```diff
@@ -71,6 +71,7 @@ show_help() {
 "" \
 "${CLI_CYAN}COMMANDS${CLI_NC}" \
 "  ${CLI_GREEN}setup [-d]${CLI_NC}                  Build and start Keycloak with OID4VCI features" \
+"  ${CLI_GREEN}compose${CLI_NC}                     Run docker compose commands (e.g., 'kc compose up -d')" \
```
I can see `(e.g., 'kc compose up -d')` do we have any `kc` command?

Awambeng reviewed (changes requested):
During the credential request (`./keycloak-ssi test preauth KMACredential` and others...), the configuration loading logs are printed multiple times within a single script run.
Consider updating the scripts so that [INFO] Loading configuration… and [INFO] Applying overrides… are logged only once per execution.

Also, could you please explain how variables can be overridden at runtime in a Docker environment with this new configuration approach? I couldn't test because of the connection.

mbunwe-victor committed:
address Bolan's comments

Awambeng committed:
Enhance Docker setup and configuration

- Update Dockerfile to install runtime dependencies and copy necessary files.
- Modify docker-compose.yml to adjust database options.
- Change Keycloak target branch in config.yaml from datev/develop to main.
- Update entrypoint script for better execution flow.
- Refactor various scripts to use PROJECT_TARGET_DIR instead of TARGET_DIR for consistency.
- Improve CLI documentation for clarity on commands.

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Awambeng committed:
Refactor Docker setup and enhance configuration management

- Clean up Dockerfile by removing unnecessary lines and improving comments.
- Update config.yaml to clarify keystore path for Docker usage.
- Modify docker-compose.yml to enable volume mapping for config.override.yaml.
- Revise docker-entrypoint.sh to improve initialization logic and error handling.
- Enhance keycloak-ssi.sh to ensure cryptographic materials are present before starting Keycloak.
- Streamline setup-kc-oid4vci.sh by delegating SSL key generation to a new helper function.
- Add ensure_keycloak_crypto_materials function in helper.sh for better cryptographic material management.
- Remove dynamic keystore path determination from import_kc_config.sh for clarity.

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Awambeng committed:
Refactor Docker and Keycloak configuration management

- Remove volume mapping for config.override.yaml in docker-compose.yml.
- Update keycloak-ssi.sh to ensure cryptographic materials are only checked during start commands.
- Enhance .env file generation by merging configuration files and ensuring only existing files are processed.
- Modify realm attributes in JSON configuration files to use environment variables for better flexibility.
- Change return statement in generate_keystore.sh for consistency in script behavior.

Signed-off-by: Awambeng <awambengrodrick@gmail.com>

Awambeng committed:
Update keystore path to default

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Awambeng committed:
Add config.override.example.yaml for Keycloak SSI configuration overrides

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Awambeng committed:
Merge pull request #190 from adorsys/improve-yaml-override-config

Improve yaml override config

mbunwe-victor committed:
update the CLI_GUIDE.md file
